flowauth/backend/flowauth/token_management.py

```diff
@@ -134,47 +102,63 @@ def add_token(server):
-    and claims is a nested dict of the form {<claim_name>:{<right_name>:<bool>}}.
+    Expects json with "name", and "roles" keys, where "name" is a string,
+    and roles is a list of role names.
 
     Responds with a json object {"token":<token_string>, "id":<token_id>}.
 
     """
-    server = Server.query.filter(Server.id == server).first_or_404()
+    server = Server.query.filter(Server.id == server_id).first_or_404()
     json = request.get_json()
 
     current_app.logger.debug("New token request", request=json)
-
     if "name" not in json:
         raise InvalidUsage("No name.", payload={"bad_field": "name"})
-    expiry = datetime.datetime.strptime(json["expiry"], "%Y-%m-%dT%H:%M:%S.%fZ")
-    lifetime = expiry - datetime.datetime.now()
-    latest_lifetime = current_user.latest_token_expiry(server)
-    if expiry > latest_lifetime:
-        raise InvalidUsage("Token lifetime too long", payload={"bad_field": "expiry"})
-    allowed_claims = current_user.allowed_claims(server)
+    if "roles" not in json:
+        raise InvalidUsage("No roles.", payload={"bad_field": "roles"})
+
+    user_roles = db.session.execute(
+        select(Role)
+        .join(User.roles)
+        .filter(User.id == current_user.id)
+        .filter(Role.server_id == server_id)
+    ).all()
+    user_roles = [row[0] for row in user_roles]
+
+    roles = []
+    for requested_role in json["roles"]:
+        if requested_role["name"] not in [ur.name for ur in user_roles]:
+            raise Unauthorized(
+                f"Role '{requested_role['name']}' is not permitted for the current user"
+            )
+        roles.append(Role.query.filter(Role.name == requested_role["name"]).first())
+    token_expiry = min(server.next_expiry(), min(rr.next_expiry() for rr in roles))
+    # The role expiry date doesn't beat the server expiry date
+    # The role longest lifetime doesn't beat the server longest lifetime
+    # If you request token with a role with a expiry past the server final expiry, then issue the token with the server's final expiry
+    # feature todo: flag this to the user
+    # This isn't about the user, so get these values from the server
 
-    current_app.logger.debug("New token request", allowed_claims=allowed_claims)
-    for claim in json["claims"]:
-        if claim not in allowed_claims:
-            raise Unauthorized(f"You do not have access to {claim} on {server.name}")
+    if token_expiry < datetime.datetime.now():
+        raise Unauthorized(f"Token for {current_user.username} expired")
 
     token_string = generate_token(
-        username=current_user.username,
         flowapi_identifier=server.name,
-        lifetime=lifetime,
-        claims=json["claims"],
+        username=current_user.username,
         private_key=current_app.config["PRIVATE_JWT_SIGNING_KEY"],
+        lifetime=token_expiry - datetime.datetime.now(),
+        roles={role.name: sorted([ss.name for ss in role.scopes]) for role in roles},
```

Comment:
```suggestion
        roles={role.name: sorted(ss.name for ss in role.scopes) for role in roles},
```

Don't need the comprehension here really.